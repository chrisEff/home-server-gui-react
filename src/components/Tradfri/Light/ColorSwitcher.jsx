'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setTradfriBulbColor} from '../../../actions'

const ColorSwitcher = ({deviceId, bulbType, color, onChange}) => {
	if (bulbType === 'rgb') {
		return (
			<div className='color-switcher'>
				<div className={color === 'red'     ? 'color red active'     : 'color red'}     onClick={() => onChange(deviceId, 'red')}><div></div></div>
				<div className={color === 'green'   ? 'color green active'   : 'color green'}   onClick={() => onChange(deviceId, 'green')}><div></div></div>
				<div className={color === 'blue'    ? 'color blue active'    : 'color blue'}    onClick={() => onChange(deviceId, 'blue')}><div></div></div>
				<div className={color === 'yellow'  ? 'color yellow active'  : 'color yellow'}  onClick={() => onChange(deviceId, 'yellow')}><div></div></div>
				<div className={color === 'pink'    ? 'color pink active'    : 'color pink'}    onClick={() => onChange(deviceId, 'pink')}><div></div></div>
				<div className={color === 'purple'  ? 'color purple active'  : 'color purple'}  onClick={() => onChange(deviceId, 'purple')}><div></div></div>
				<div className={color === 'warm'    ? 'color warm active'    : 'color warm'}    onClick={() => onChange(deviceId, 'warm')}><div></div></div>
				<div className={color === 'neutral' ? 'color neutral active' : 'color neutral'} onClick={() => onChange(deviceId, 'neutral')}><div></div></div>
				<div className={color === 'cold'    ? 'color cold active'    : 'color cold'}    onClick={() => onChange(deviceId, 'cold')}><div></div></div>
			</div>
		)
	}
	if (bulbType === 'white-spectrum') {
		return (
			<div className='color-switcher spectrum'>
				<div className={color === 'warm'    ? 'color warm active'    : 'color warm'}    onClick={() => onChange(deviceId, 'warm')}><div></div></div>
				<div className={color === 'neutral' ? 'color neutral active' : 'color neutral'} onClick={() => onChange(deviceId, 'neutral')}><div></div></div>
				<div className={color === 'cold'    ? 'color cold active'    : 'color cold'}    onClick={() => onChange(deviceId, 'cold')}><div></div></div>
			</div>
		)
	}
	if (bulbType === 'white') {
		return (
			<div className='color-switcher white'>
				<div className='color neutral active'><div></div></div>
			</div>
		)
	}
	return <div/>
}

ColorSwitcher.propTypes = {
	deviceId: PropTypes.number.isRequired,
	bulbType: PropTypes.string.isRequired,
	color: PropTypes.string,
	onchange: PropTypes.func,
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, color) => dispatch(setTradfriBulbColor(id, color)),
})

export default connect(mapStateToProps, mapDispatchToProps)(ColorSwitcher)
export {ColorSwitcher}
