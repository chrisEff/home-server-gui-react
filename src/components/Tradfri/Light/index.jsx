'use strict'

const homeServerApi = require('../../../homeServerApi')

const React = require('react')
const PropTypes = require('prop-types')

const LightTitle = require('./LightTitle')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {

	static propTypes = {
		bulb: PropTypes.object.isRequired,
	}

	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<LightTitle title={bulb.name} onChange={this.updateName} />

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateName = async (name) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/name/${name}`)

		bulb.name = name
		this.setState({bulb})
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/state/${state}`)

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/brightness/${brightness}`)

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await homeServerApi.put(`/tradfri/device/${bulb.id}/color/${color}`)

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
