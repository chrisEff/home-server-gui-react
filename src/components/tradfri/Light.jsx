'use strict'

const api = require('../../../config').api
api.url = `${api.protocol}://${api.host}:${api.port}`

const React = require('react')
const PropTypes = require('prop-types')

const LightTitle = require('./LightTitle')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {

	static propTypes = {
		bulb: PropTypes.object.isRequired,
	}

	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<LightTitle title={bulb.name} onChange={this.updateName} />

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateName = async (name) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/name/${name}?key=${api.key}`, {method: 'PUT'})

		bulb.name = name
		this.setState({bulb})
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/state/${state}?key=${api.key}`, {method: 'PUT'})

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${api.key}`, {method: 'PUT'})

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(`${api.url}/tradfri/device/${bulb.id}/color/${color}?key=${api.key}`, {method: 'PUT'})

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
