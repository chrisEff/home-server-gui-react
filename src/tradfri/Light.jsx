'use strict'

const config = require('../../config')

const React = require('react')
const BrightnessSlider = require('./BrightnessSlider')
const ColorSwitcher = require('./ColorSwitcher')
const LightSwitch = require('./LightSwitch')

class Light extends React.Component {
	constructor (props) {
		super(props)
		this.state = {bulb: props.bulb}
	}

	render () {
		const bulb = this.state.bulb;
		const className = bulb.state ? 'light on' : 'light'
		return (
			<div className={className + ' ' + bulb.color}>
				<span className={bulb.name.length > 15 ? 'name long' : 'name'}>{bulb.name}</span>

				<LightSwitch state={bulb.state} onchange={this.updateState}/>
				<BrightnessSlider brightness={bulb.brightness} onchange={this.updateBrightness}/>
				<ColorSwitcher bulbType={bulb.bulbType} color={bulb.color} onchange={this.updateColor}/>
			</div>
		)
	}

	updateState = async (state) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/state/${state}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.state = state
		this.setState({bulb})
	}

	updateBrightness = async (brightness) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/brightness/${brightness}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.brightness = brightness
		this.setState({bulb})
	}

	updateColor = async (color) => {
		const bulb = this.state.bulb
		await fetch(
			`${config.api.protocol}://${config.api.host}:${config.api.port}/tradfri/device/${bulb.id}/color/${color}?key=${config.api.key}`,
			{method: 'PUT'}
		)

		bulb.color = color
		this.setState({bulb})
	}
}

module.exports = Light
