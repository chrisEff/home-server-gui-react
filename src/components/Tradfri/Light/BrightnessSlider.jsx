'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { StyleSheet, css } from 'aphrodite'

import { setDeviceBrightness } from '@/actions/tradfri'

const BrightnessSlider = ({ deviceId, brightness, onChange }) =>
	<input
		className={css([styles.brightnessSlider])}
		type="range"
		min="0"
		max="254"
		defaultValue={brightness}
		onMouseUp={(e) => onChange(deviceId, parseInt(e.target.value))}
	/>

BrightnessSlider.propTypes = {
	deviceId: PropTypes.number.isRequired,
	brightness: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const styles = StyleSheet.create({
	brightnessSlider: {
		cursor: 'pointer',
		margin: '20px 0',
		width: '120px',
	},
})

const mapStateToProps = (state, ownProps) => ({})
const mapDispatchToProps = (dispatch) => ({
	onChange: (id, brightness) => dispatch(setDeviceBrightness(id, brightness)),
})

export default connect(mapStateToProps, mapDispatchToProps)(BrightnessSlider)
export { BrightnessSlider }
