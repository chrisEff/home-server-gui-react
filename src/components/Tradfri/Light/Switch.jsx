'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {setDeviceState} from '@/actions/tradfri'

let Switch = ({deviceId, state, onChange}) =>
	<div style={state ? {...styles.bulb, ...styles.on} : styles.bulb} className='bulb' onClick={() => onChange(deviceId, state ? 0 : 1)}>ðŸ’¡</div>


Switch.propTypes = {
	deviceId: PropTypes.number.isRequired,
	state: PropTypes.number.isRequired,
	onchange: PropTypes.func,
}

const styles = {
	bulb: {
		filter: 'grayscale(100%) blur(5px)',
		fontSize: 80,
		paddingTop: '15px',
		textAlign: 'center',
		userSelect: 'none',
	},
	on: {
		filter: 'none',
	},
}

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, state) => dispatch(setDeviceState(id, state)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Switch)
export {Switch}
