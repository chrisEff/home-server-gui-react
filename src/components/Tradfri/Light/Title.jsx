'use strict'

import React, {useRef, useState} from 'react'
import PropTypes from 'prop-types'
import {connect} from 'react-redux'
import {StyleSheet, css} from 'aphrodite'

import {setDeviceName} from '@/actions/tradfri'

const Title = ({deviceId, title, onChange}) => {
	const [editMode, setEditMode] = useState(false)
	const titleInput = useRef(title)

	if (!editMode) {
		return (
			<span className={
				css(title.length > 15 ? [styles.title, styles.long] : [styles.title])
			} onClick={() => setEditMode(true)}>
				{title}
			</span>
		)
	} else {
		return (
			<div>
				<input type='text' ref={titleInput} style={{width: 110}} defaultValue={title} />
				<button onClick={() => { setEditMode(false); onChange(deviceId, titleInput.current.value) }}>OK</button>
			</div>
		)
	}
}

Title.propTypes = {
	deviceId: PropTypes.number.isRequired,
	title: PropTypes.string.isRequired,
	onChange: PropTypes.func,
}

const styles = StyleSheet.create({
	title: {
		':hover': {
			':after': {
				content: "' ✏️'",
				fontSize: '0.9em',
			},
		},
		display: 'block',
		height: '1.6em',
	},
	long: {
		fontSize: '0.8em',
		height: '2em',
	},
})

const mapStateToProps = (state, ownProps) => ({
})
const mapDispatchToProps = dispatch => ({
	onChange: (id, name) => dispatch(setDeviceName(id, name)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Title)
export {Title}
