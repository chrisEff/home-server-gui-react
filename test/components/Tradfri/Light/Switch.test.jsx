'use strict'

import React from 'react'
import { Switch } from '@/components/Tradfri/Light/Switch'

import { shallow } from 'enzyme'

describe('Switch', () => {
	it('should render correctly', () => {
		const wrapper = shallow(<Switch deviceId={1} state={1} />)
		expect(wrapper).toMatchSnapshot()
	})
})
