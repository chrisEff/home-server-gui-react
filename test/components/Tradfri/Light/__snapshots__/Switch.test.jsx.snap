// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render correctly 1`] = `
<div
  className="bulb"
  onClick={[Function]}
  style={
    Object {
      "filter": "none",
      "fontSize": 80,
      "paddingTop": "15px",
      "textAlign": "center",
      "userSelect": "none",
    }
  }
>
  ðŸ’¡
</div>
`;
